Feature/prepare for exe

han-thony opened the pull request:
Allow for CAN to be configured from GUI without needing to modify code. Prepare for windows exe build. Almost all existing Pylance errors removed, no new pylance errors introduced. I also added logging. 

We've successfully launched the exe on a windows machine. Here are the builds: https://github.com/envgoinc/can_testbench/actions

### Changelog:

Connection settings
- UI for selecting bus type and connection options
- Options dynamically enabled/disabled based on selected bus type
- Options will be saved upon successful connection in config file, can simply hit connect button if desired config is unchanged from most recent use
- DBC files with initial values outside of the min max such as "Signal Not Available" no longer fail to open.
- Connect button disabled and red DBC path when no valid DBC is loaded.

Tab system
- More than one CAN connection can be opened at a time as long as the channel is unique
- If a channel is reopened, old tab and CAN connection will be closed
- Tabs can now be manually closed, will close the connection and both TX and RX tabs
- Graph window can now be manually closed and will correctly update the message table
- Fixed graph window remaining open when application or parent tab is closed
- Can connections shutdown properly when application or parent tab is closed

Miscellaneous
- Added logging system using the canutil *.log ASCII format.
- Named values will now have their names displayed in the table
- Pop up dialog for error messages with dynamic description
- Connection details at the top right on every tab so you don't mix them up
- Refactored code to improve maintainability
- Improved can_send.py script for more thorough testing.
- Fixed bug where once you stop sending a message, you can't send it again until you change frequency
- Github action to automate MacOS, Windows and Ubuntu builds
- A few other small bug fixes

Known issues
- If initial value is out of range such as "Signal Not Available", you cannot change back to it after switching off.

han-thony commented on can_testbench.py:
```diff
@@ -1,23 +1,18 @@
 from cantools.database.namedsignalvalue import NamedSignalValue
-import can
-import os
+import can as pycan
```
Changed imports to avoid importing classes instead of modules, makes it less confusing when there are two can libraries that implement some of the same classes.

jmailloux commented on can_testbench.py:
I get the following error when trying to run the script:
can_testbench/can_testbench.py
INFO:root:3.12.2 (main, Feb  6 2024, 20:19:44) [Clang 15.0.0 (clang-1500.1.0.2.5)]
Traceback (most recent call last):
  File "/Users/jerrymailloux/dev/can_testbench/can_testbench.py", line 966, in <module>
    mainApp = MainApp()
              ^^^^^^^^^
  File "/Users/jerrymailloux/dev/can_testbench/can_testbench.py", line 870, in __init__
    self.initUI()
  File "/Users/jerrymailloux/dev/can_testbench/can_testbench.py", line 958, in initUI
    self.setupLaunchTab()
  File "/Users/jerrymailloux/dev/can_testbench/can_testbench.py", line 903, in setupLaunchTab
    tabBar.tabButton(0, QTabBar.RightSide).deleteLater()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'deleteLater'

jmailloux replied:
Also, would be good if you can clean up the pylance issues.

han-thony replied:
All pylance errors from my changes have been removed, I also fixed a lot of existing pylance errors.

jmailloux reviewed (changes requested):
In general, I like the direction you are going.  You've made some good refactoring changes.  I have to take a closer look at some of the changes are some are quite big.
You can see my initial comments.  I will take a closer look this weekend and give you more comments.

han-thony commented on can_testbench.py:
```diff
@@ -591,14 +782,37 @@ def onSignalValueChanged(self, msg: DbcMessage, row: int, value: object, graph:
+            message = DbcMessage(message=msg, signals=[])
+            for sig in msg.signals:
+                if isinstance(sig.initial, NamedSignalValue):
+                    value = sig.initial.name
```
sig.initial can't be none if sig.initial is an instance of NamedSignalValue.

han-thony commented on can_testbench.py:
```diff
@@ -352,6 +364,216 @@ def closeEvent(self, event):
+
+    """ 
+    def __init__(self):
+        super().__init__()
```
Oops, accidentally put that there

han-thony commented on can_testbench.py:
```diff
@@ -352,6 +364,216 @@ def closeEvent(self, event):
+            self.options[self.index()]['channel'] = channel
+        
+    def changeBitrate(self, index: int):
+        if 'bitrate' in self.options[self.index()]:
```
You're right, it makes more sense to take the bitrate and just check to make sure it's in the tuple instead of taking an index.

han-thony replied:
I used index originally because the SLCAN class has a list of valid bitrates that the bitrate must be on or it will error. I instinctively wanted to avoid a linear search and guarantee the value would always be valid so I used the index. However, A) I'll just use an enum to check if the bitrate is valid and B) The performance hit is negligible even if I didn't.

han-thony commented on can_testbench.py:
```diff
@@ -30,27 +28,34 @@
     graphValues (deque): A deque of max 100 values that will be graphed.
     Values are the latest values received
-    graph (bool): Whether or not the signal should be graphed.
+    graphed (bool): Whether or not the signal should be graphed.
```
Previous naming implies graph is an object instead of a bool, I think there was actually an incorrect existing type hint that treated DbcSignal.graph as an object.

han-thony commented on can_testbench.py:
```diff
@@ -146,21 +155,28 @@ def stop(self, msg):
     """
-    SignalValueChanged = Signal(DbcMessage, int, object, object)
+    signalValueChanged = QtCore.Signal(DbcMessage, int, float)
+    signalGraphedChanged = QtCore.Signal(DbcMessage, int, bool, object)
```
I split this into two signals because the original slot felt like it wanted to be two different functions for two different situations. This signal was only emitted twice, once with the first object as null and once with the second object as null.

han-thony commented on can_testbench.py:
```diff
@@ -141,26 +150,35 @@ def sendCanMessage(self, msg, frequency=0):
                 else:
                     sendDetails['task'].start()
+            else:
+                sendDetails['task'].start()
```
This fixes an existing bug where once you stop sending, you can't send that message again unless you change the frequency.

han-thony commented on can_testbench.py:
```diff
@@ -352,6 +364,216 @@ def closeEvent(self, event):
+        super().__init__()
+        self.config = configparser.ConfigParser()
+        self.scriptDir = path.dirname(path.abspath(__file__))
+        self.configFile = path.join(self.scriptDir, 'can_testbench.ini')
```
Shelving for now

han-thony commented on can_testbench.py:
```diff
@@ -1,23 +1,27 @@
+# nuitka-project: --disable-console
+# nuitka-project: --standalone
+# nuitka-project: --include-module=can.interfaces.slcan
+# nuitka-project: --include-module=can.interfaces.udp_multicast
```
Have to manually add any new can.interfaces we implement here because python-can uses dynamic imports for the interfaces which aren't automatically detected by Nuitka

jmailloux commented on can_testbench.py:
```diff
@@ -340,18 +377,248 @@ def updatePlot(self):
-                permitClose = False
-
+        
         if permitClose:
```
Why keep permitClose if it is now always true?

han-thony replied:
At the time I was considering if I might want to add some sort of condition, I will remove it.

jmailloux commented on can_testbench.py:
```diff
@@ -500,144 +777,279 @@ class RxMessageLayout(MessageLayout):
-
+        self.tabWidget.addTab(tab, 'CAN Config')
+        tabBar = self.tabWidget.tabBar()
+        tabBar.tabButton(0, QTabBar.ButtonPosition.RightSide).resize(0, 0)
```
This is failing for me when I try to run it.
  File "/Users/jerrymailloux/dev/can_testbench/can_testbench.py", line 976, in setupLaunchTab
    tabBar.tabButton(0, QTabBar.ButtonPosition.RightSide).resize(0, 0)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'resize'

han-thony replied:
Should be fixed now, I now check both sides and check that the button exists before calling resize.

jmailloux reviewed (changes requested):
I need to be able to run this and test it before you merge.
I like the changes.  Look forward to trying it out!

han-thony commented on can_testbench.py:
```diff
@@ -1,23 +1,27 @@
+import enum
+from cantools import database
+from cantools.database.can import signal
+from cantools.database.namedsignalvalue import NamedSignalValue
```
"database" and "signal" are modules which have lowercase names, NamedSignalValue is a specific class. I like to avoid importing specific classes to decrease risk of name collisions and because programmatically the whole module gets processed anyways. The NamedSignalValue import was merged from main so I wasn't sure if I should change it.

han-thony commented:
Fixed the issue with SLCAN, it's because I forgot to sanitize the channel string when using it in the name for the logfile. SLCAN channel names have the form "/dev/tty...". Also fixed existing warning when no signal.graphValues exist for a graphed message.
